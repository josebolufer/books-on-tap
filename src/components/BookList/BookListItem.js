import { Link as RouterLink } from 'react-router-dom'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import CardMedia from '@material-ui/core/CardMedia'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'

const useStyles = makeStyles((theme) => ({
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  cardMedia: {
    maxHeight: '250px',
  },
  cardContent: {
    flexGrow: 1,
  },
}))

const BookListItem = ({ book }) => {
  const classes = useStyles()
  const { id, thumbnail, title, price, author } = book

  return (
    <Grid data-testid='book-list-item' item xs={12} sm={6} md={4}>
      <Card raised className={classes.card}>
        <CardMedia
          className={classes.cardMedia}
          src={thumbnail}
          title={title}
          alt={title}
          component='img'
        />
        <CardContent className={classes.cardContent}>
          <Typography
            gutterBottom
            variant='body1'
            component='h2'
            color='primary'
          >
            {title}
          </Typography>
          <Typography
            variant='body2'
            gutterBottom
          >{`By ${author.firstName} ${author.lastName}`}</Typography>
          <Typography variant='body2'>Â£{price.toFixed(2)}</Typography>
        </CardContent>
        <CardActions>
          <Button
            variant='outlined'
            size='small'
            color='primary'
            component={RouterLink}
            to={`/book/${id}`}
          >
            View
          </Button>
        </CardActions>
      </Card>
    </Grid>
  )
}

BookListItem.propTypes = {
  book: PropTypes.object.isRequired,
}

export default BookListItem
