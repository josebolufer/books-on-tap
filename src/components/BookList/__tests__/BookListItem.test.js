import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import BookListItem from '../BookListItem'

const mockedBook = {
  author: { firstName: 'Jose', lastName: 'Bolufer' },
  id: '1',
  description: 'The story about a boy that grew up in London',
  isbn: 'asdfghghghghg',
  price: 10,
  stockAmount: 1,
  thumbnail: 'Image 1',
  title: 'A live in London',
}

describe('@renders', () => {
  it('in default state', () => {
    const { asFragment } = render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render the book image', () => {
    render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })

    expect(screen.getByAltText(mockedBook.title)).toHaveAttribute(
      'src',
      mockedBook.thumbnail
    )
  })

  it('should render the book title', () => {
    render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })

    expect(screen.getByText(mockedBook.title)).toBeInTheDocument()
  })

  it('should render the book author', () => {
    render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })

    expect(
      screen.getByText(
        `By ${mockedBook.author.firstName} ${mockedBook.author.lastName}`
      )
    ).toBeInTheDocument()
  })

  it('should render the price of the book', () => {
    render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })

    expect(
      screen.getByText(`Â£${mockedBook.price.toFixed(2)}`)
    ).toBeInTheDocument()
  })

  it('should render the view button', () => {
    render(<BookListItem book={mockedBook} />, {
      wrapper: MemoryRouter,
    })

    expect(screen.getByText('View').closest('a')).toHaveAttribute(
      'href',
      `/book/${mockedBook.id}`
    )
  })
})
