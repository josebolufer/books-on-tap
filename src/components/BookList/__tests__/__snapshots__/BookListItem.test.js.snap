// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@renders in default state 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    data-testid="book-list-item"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation8 MuiPaper-rounded"
    >
      <img
        alt="A live in London"
        class="MuiCardMedia-root makeStyles-cardMedia-2 MuiCardMedia-media MuiCardMedia-img"
        src="Image 1"
        title="A live in London"
      />
      <div
        class="MuiCardContent-root makeStyles-cardContent-3"
      >
        <h2
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorPrimary MuiTypography-gutterBottom"
        >
          A live in London
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
        >
          By Jose Bolufer
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Â£10.00
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
          href="/book/1"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            View
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
